clients/subscriberlite.go

```diff
@@ -0,0 +1,170 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clients
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"cloud.google.com/go/pubsub"
+	"cloud.google.com/go/pubsublite/pscompat"
+	"github.com/conduitio-labs/conduit-connector-gcp-pubsub/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/gammazero/deque"
+	"google.golang.org/api/option"
+)
+
+const subscriptionPathFmt = "projects/%s/locations/%s/subscriptions/%s"
+
+// SubscriberLite represents a struct with a GCP Pub/Sub Lite client,
+// queues for messages, channels and a function to cancel the context.
+type SubscriberLite struct {
+	subscriber *pscompat.SubscriberClient
+
+	lock       sync.Mutex
+	msgDeque   *deque.Deque[*pubsub.Message]
+	ackDeque   *deque.Deque[*pubsub.Message]
+	errorCh    chan error
+	canceledCh chan struct{}
+	ctxCancel  context.CancelFunc
+}
+
+// NewSubscriberLite initializes a new subscriber client of GCP Pub/Sub Lite
+// and starts receiving a messages to the queue.
+func NewSubscriberLite(ctx context.Context, cfg config.Source) (*SubscriberLite, error) {
+	credential, err := cfg.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	subscriptionPath := fmt.Sprintf(subscriptionPathFmt, cfg.ProjectID, cfg.Location, cfg.SubscriptionID)
+
+	subscriber, err := pscompat.NewSubscriberClientWithSettings(ctx, subscriptionPath, pscompat.ReceiveSettings{
+		NackHandler: func(message *pubsub.Message) error {
+			sdk.Logger(ctx).Info().Msgf("message wasn't ack: %q", message.ID)
+
+			return nil
+		},
+	}, option.WithCredentialsJSON(credential))
+	if err != nil {
+		return nil, fmt.Errorf("new subscriber client: %w", err)
+	}
+
+	cctx, cancel := context.WithCancel(ctx)
```

Comment:
Except for this part, it looks like `subscriber` and `subscriberlite` are pretty much identical?

Reply:
Correct. Also, `Stop` methods are pretty similar.
I've decided to solve it with struct embedding.